Refactor text editing replacement mechanism

rr-codes opened the pull request:
#### 3a34f4ad4c4354c15ae7e832ff26e5fdf5222533
<pre>
Refactor text editing replacement mechanism
<a href="https://bugs.webkit.org/show_bug.cgi?id=268214">https://bugs.webkit.org/show_bug.cgi?id=268214</a>
<a href="https://rdar.apple.com/120563444">rdar://120563444</a>

Reviewed by Aditya Keerthi.

Introduce a more unified way of handling text edits and replacements.

* Source/WebCore/dom/DocumentMarker.h:
(WebCore::DocumentMarker::allMarkers):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::rangeByExtendingCurrentSelection const):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::markerTypeFrom):
(WebCore::Internals::hasUnifiedTextReplacementMarker):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebTextReplacementData.h: Added.
* Source/WebKit/Shared/WebUnifiedTextReplacementContextData.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didBeginTextReplacementSession):
(WebKit::WebPageProxy::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::rangeSelectionContext const):
(WebKit::WebPage::didBeginTextReplacementSession):
(WebKit::WebPage::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp: Added.
(WebKit::UnifiedTextReplacementController::UnifiedTextReplacementController):
(WebKit::UnifiedTextReplacementController::didBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h: Added.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_unifiedTextReplacementController):
(WebKit::m_historyItemClient): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedTextReplacement.mm: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:

Canonical link: <a href="https://commits.webkit.org/273687@main">https://commits.webkit.org/273687@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad9a9b2c5e73cd67794643374623350d5c06de5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/36381 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/15333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/38600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/32691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/37610 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17782 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/12365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/39093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/36941 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/12945 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/38600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39093 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/38600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/33010 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/38600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/40338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/11584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/12365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40338 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/38600 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4713 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/12445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -831,6 +832,24 @@
 }
 #endif
 
+SimpleRange WebPage::getRangeSelectionContext() const
```
nit: remove get from function name.

pxlcoder commented on Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
```diff
@@ -369,6 +369,13 @@
 /* End PBXBuildRule section */
 
 /* Begin PBXContainerItemProxy section */
+		074AD71C2B61795800FFA67B /* PBXContainerItemProxy */ = {
```
What's going on in this file 😅

rr-codes replied:
literally no idea whatsoever, I’ll try to put it back

pxlcoder commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2626,6 +2626,24 @@ void FrameSelection::showTreeForThis() const
 
 #endif
 
+#if PLATFORM(COCOA)
```
Let's remove the `#if` since there's nothing Cocoa-specific here.

pxlcoder commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -19006,7 +19010,6 @@
 			isa = PBXTargetDependency;
 			platformFilters = (
 				ios,
-				xros,
```
More strange `xcodeproj` changes...

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
```diff
@@ -831,6 +832,24 @@
+SimpleRange WebPage::getRangeSelectionContext() const
+{
+    Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
+    return frame->selection().rangeByExtendingCurrentSelection(TextGranularity::ParagraphGranularity).value();
```
Why is it okay to avoid null-checking the optional?

rr-codes replied:
```
Ref frame = CheckedRef(m_page->focusController())->focusedOrMainFrame();
```

is present in dozens of other places, and none of those places do a null check

rr-codes replied:
OH

rr-codes replied:
the optional

rr-codes replied:
oops

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -0,0 +1,84 @@
+    auto contextRange = m_webPage->getRangeSelectionContext();
+    auto liveRange = createLiveRange(contextRange);
+
+    RELEASE_ASSERT(!m_sessionRanges.contains(uuid));
```
`ASSERT` rather than `RELEASE_ASSERT`.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -0,0 +1,84 @@
+    if (!m_webPage)
+        return;
+
+    RELEASE_ASSERT(m_sessionRanges.contains(uuid));
```
`ASSERT` rather than `RELEASE_ASSERT`.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -0,0 +1,84 @@
+    auto liveRange = m_sessionRanges.get(uuid);
+    auto sessionRange = makeSimpleRange(liveRange);
+    if (!sessionRange) {
+        RELEASE_ASSERT_NOT_REACHED();
```
I'm not convinced an assertion is necessary here.

pxlcoder commented on Source/WebCore/page/cocoa/WKTextReplacementData.h:
```diff
@@ -0,0 +1,52 @@
+
+namespace WebCore {
+
+struct WKTextReplacementData {
```
This should be `WebTextReplacementData`, same comment applies to other uses of `WK` in this patch.

This is in WebCore, so it shouldn't know about WebKit, which is higher in the stack.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp:
```diff
@@ -0,0 +1,84 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
```
This is already correct. The correct license to use is documented here: https://webkit.org/licensing-webkit/

pxlcoder commented on Source/WebKit/Sources.txt:
```diff
@@ -836,6 +836,7 @@ WebProcess/WebCoreSupport/WebWorkerClient.cpp
 WebProcess/WebPage/DrawingArea.cpp
 WebProcess/WebPage/EventDispatcher.cpp
 WebProcess/WebPage/FindController.cpp
+WebProcess/WebPage/UnifiedTextReplacementController.cpp
```
Sort order.

aprotyas reviewed (commented):
LGTM

rr-codes committed:
Refactor text editing replacement mechanism
https://bugs.webkit.org/show_bug.cgi?id=268214
rdar://120563444

Reviewed by Aditya Keerthi.

Introduce a more unified way of handling text edits and replacements.

* Source/WebCore/dom/DocumentMarker.h:
(WebCore::DocumentMarker::allMarkers):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::rangeByExtendingCurrentSelection const):
* Source/WebCore/editing/FrameSelection.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::markerTypeFrom):
(WebCore::Internals::hasUnifiedTextReplacementMarker):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/Logging.h:
* Source/WebKit/Shared/UnifiedTextReplacement.serialization.in: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/Shared/WebTextReplacementData.h: Added.
* Source/WebKit/Shared/WebUnifiedTextReplacementContextData.h: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::didBeginTextReplacementSession):
(WebKit::WebPageProxy::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::rangeSelectionContext const):
(WebKit::WebPage::didBeginTextReplacementSession):
(WebKit::WebPage::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.cpp: Added.
(WebKit::UnifiedTextReplacementController::UnifiedTextReplacementController):
(WebKit::UnifiedTextReplacementController::didBeginTextReplacementSession):
(WebKit::UnifiedTextReplacementController::textReplacementSessionDidReceiveReplacements):
* Source/WebKit/WebProcess/WebPage/UnifiedTextReplacementController.h: Added.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_unifiedTextReplacementController):
(WebKit::m_historyItemClient): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedTextReplacement.mm: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIScripting.mm:

Canonical link: https://commits.webkit.org/273687@main
